### ISSUE TITLE
[DEPR]: Code Owner Monitoring

### ISSUE THREAD

**Issue by robrap** [openedx/edx-django-utils#469]
### Proposal Date

2024-12-11

### Target Ticket Acceptance Date

2024-12-23

### Earliest Open edX Named Release Without This Functionality

Teak - 2025-04

### Rationale

The [code_owner monitoring code](https://github.com/openedx/edx-django-utils/tree/master/edx_django_utils/monitoring/internal/code_owner) in was introduced to help 2U monitor the large monolith (LMS), and attribute different parts of the app to different teams. This DEPR assumes that this is 2U-specific code, and that no other orgs or teams are using this functionality for monitoring purposes.

Slightly related, the New Relic search script was likely only used by 2U. This could be removed if no one is using it.

### Removal

What would be removed is https://github.com/openedx/edx-django-utils/tree/master/edx_django_utils/monitoring/internal/code_owner, and all associated code, config, and documentation.

Slightly related, we could also remove [the new_relic_search.py script](https://github.com/openedx/edx-django-utils/blob/18cf72da7503eb283447f631d54292e5111b56b5/edx_django_utils/monitoring/scripts/new_relic_search.py) and supporting documentation. There would be no replacement, other than simply making a copy of the script as-needed.

### Replacement

- It is possible that no one in the community needs a replacement.
- For 2U, the functionality was replaced by a plugin that adds instrumentation in two ways:
  1. For Django requests, instrumentation was added to the plugin using new signals that were added to edx-django-utils MonitoringSupportMiddleware, as documented in https://github.com/openedx/edx-django-utils/pull/467.
  2. For Celery tasks, 2U is using a Datadog span processor to add instrumentation.

For 2U, the span processor replaces the need for the custom decorator that was added to all Open edX celery tasks, and documented in the ADR: [0003-code-owner-for-celery-tasks](https://github.com/openedx/edx-django-utils/blob/65b138f55f13e1935f8e9155e6f443e020ec2ac1/edx_django_utils/monitoring/docs/decisions/0003-code-owner-for-celery-tasks.rst).

If we choose to remove this decorator, there is no real replacement if your monitoring tool does not have some way to hook into these tasks. New Relic had no such hook in the past.

### Deprecation

- It could simply be removed if no one is using it.
- The decorator for celery tasks is slightly more complicated, because it has references throughout the codebase. This might get marked as deprecated before removal.

### Additional Info

If no one is using this except 2U, it could be removed shortly after acceptance, since 2U has nearly completed the migration to a plugin.

### Task List

- [ ] Remove all uses of the celery task decorator for code owner monitoring.
- [ ] Remove all non-celery task code owner code and docs.
- [ ] Remove the remaining of the code owner monitoring code.
- [ ] Remove the new_relic_search script and docs if that is agreeable.

**Comment by robrap** [comment]
Announced here: https://discuss.openedx.org/t/deprecation-removal-code-owner-monitoring-edx-django-utils-469/14554

**Comment by dianakhuang** [comment]
A team at 2U has this on backlog, but it's low priority.

**Comment by rahulkanneri-2u** [comment]
@robrap 

Status update:

- #547 is replaced by #551 (open, awaiting review)
- [#38967](https://github.com/openedx/openedx-platform/pull/38967) is replaced by [#39047](https://github.com/openedx/openedx-platform/pull/39047) (open, awaiting review)

Could someone with edit access update the task list above to point at #551 and [#39047](https://github.com/openedx/openedx-platform/pull/39047)?

**Comment by rahulkanneri-2u** [comment]
Progress update:

**openedx-platform** (Celery `set_code_owner_attribute` decorator removal):
- Merged: openedx/openedx-platform#39047
- Follow-up fix merged: openedx/openedx-platform#39096 restores the `0001-monitoring-by-code-owner.rst` ADR (accidentally deleted in #39047) with its Status updated to Superseded, pointing back to this issue.

**edx-django-utils** (non-celery code owner monitoring cleanup):
- Closed without merging: openedx/edx-django-utils#551
- Reason: removing `CodeOwnerMonitoringMiddleware` is a breaking change for downstream consumers. Doing this safely would require adding deprecation warnings first (and updating the changelog), which we're intentionally not taking on right now to keep this timeboxed. The celery decorator (`set_code_owner_attribute`) and its helpers in `utils.py` are left in place as still-in-use and out of scope for now.

**Comment by robrap** [comment]
@feanil @kdmccormick: This is back on hold.
- I imagine deprecation warnings need extra work to ensure they don’t repeat.
- in the future, could we go directly to a breaking change in edx-django-utils and use the major version to communicate the breakage without warnings and waiting a named release?

**Comment by feanil** [comment]
Given that no one is using this, I'm fine with it getting dropped via a major version bump when someone can get back to it.  It can move directly to "Breaking changes unblocked" at that point.

**Comment by robrap** [comment]
Thanks @feanil. I clarified the task list.

And if 2U is ever able to do this work, we also need to [clean-up these 2U search results](https://github.com/search?q=org%3Aedx+%28CodeOwnerMonitoringMiddleware+OR+get_code_owner_from_module+OR+set_code_owner_attribute+OR+set_code_owner_attribute_from_module%29+-repo%3Aedx%2Fedx-platform+-repo%3Aedx%2Fedx-arch-experiments++NOT+is%3Aarchived&type=code).
